client/src/main/java/io/pravega/client/security/auth/JwtTokenHandlingStrategy.java

```diff
@@ -0,0 +1,139 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.security.auth;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.gson.Gson;
+import io.pravega.client.stream.impl.Controller;
+import io.pravega.common.Exceptions;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.common.util.ConfigurationOptionsExtractor;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import java.time.Instant;
+import java.util.Base64;
+import java.util.concurrent.atomic.AtomicReference;
+
+@Slf4j
+public class JwtTokenHandlingStrategy implements DelegationTokenHandlingStrategy {
+
+    /**
+     * Represents the threshold for triggering delegation token refresh.
+     */
+    @VisibleForTesting
+    static final int DEFAULT_REFRESH_THRESHOLD = 5;
+
+    @Getter(AccessLevel.PROTECTED)
+    private static final int REFRESH_THRESHOLD;
+
+    /**
+     * The Controller gRPC client used for interacting with the server.
+     */
+    private final Controller controllerClient;
+
+    private final String scopeName;
+
+    private final String streamName;
+
+    @Getter(AccessLevel.PROTECTED)
+    private AtomicReference<DelegationToken> delegationToken = new AtomicReference<>();
+
+    static {
+        REFRESH_THRESHOLD = ConfigurationOptionsExtractor.extractInt(
+                "pravega.client.auth.token-refresh.threshold",
+                "pravega_client_auth_token-refresh.threshold",
+                DEFAULT_REFRESH_THRESHOLD);
+    }
+
+    JwtTokenHandlingStrategy(Controller controllerClient, String scopeName, String streamName) {
+        Exceptions.checkNotNullOrEmpty(scopeName, "scopeName");
+        Preconditions.checkNotNull(controllerClient, "controllerClient is null");
+        Exceptions.checkNotNullOrEmpty(streamName, "streamName");
+
+        this.scopeName = scopeName;
+        this.streamName = streamName;
+        this.controllerClient = controllerClient;
+    }
+
+    public JwtTokenHandlingStrategy(String token, Controller controllerClient, String scopeName, String streamName) {
+        Exceptions.checkNotNullOrEmpty(token, "token");
+        Exceptions.checkNotNullOrEmpty(scopeName, "scopeName");
+        Preconditions.checkNotNull(controllerClient, "controllerClient is null");
+        Exceptions.checkNotNullOrEmpty(streamName, "streamName");
+
+        Long expTime = extractExpirationTime(token);
+        delegationToken.set(new DelegationToken(token, expTime));
+        this.scopeName = scopeName;
+        this.streamName = streamName;
+        this.controllerClient = controllerClient;
+    }
+
+    @Override
+    public String retrieveToken() {
+        if (isTokenNearingExpiry()) {
+            log.info("Token is nearing expiry, so refreshing it");
+            return refreshToken();
```

Comment:
This does not guard against many calls to get the token coming in while it is being refreshed. We only want to update the token once if it is near expiry. There are various classes in `common` which might help with this. Perhaps a simpler solution of backgrounding the refresh, and using the current token so that calls in the meantime are not blocked.

Reply:
I wanted to avoid the use of locks, even though in some cases the code will indeed lead to the delegation token wrapped by AtomicReference getting replaced multiple times unnecessarily. A background refresh will also require locks. What do you say?

Reply:
Then you should try to use optimistic concurrency. Make use of the primitives offered by `AtomicReference` to do compare-and-set to guard against updating it to a stale value.

Reply:
@andreipaduroiu The problem is not the re-writing of the value, the problem is fetching it multiple times. 

@ravisharda How about the following, have an atomic reference and an atomic boolean to indicate if it is being fetched. If the token is actually expired: Fetch the token and wait for the new value. If the token is near but not yet expired, and the AtomicBoolean is false, set it to true (with compare and set), then trigger a background fetch of the token and get a new token, but return to the caller the one already in memory. 

This way in the happy case where the token gets refreshed before expiry only one caller ever refreshes the token. And we avoid the need for any locking

Reply:
Thanks @tkaitchuck and @andreipaduroiu. I've modified the code as suggested by both of you in the latest commit.  Please take a look and see if it makes sense.